packages/common/src/uleb128.ts

```diff
@@ -1,19 +1,57 @@
 import { Buffer } from 'buffer/';
+import { unwrap } from './util';
 
 /**
- * Decodes an unsigned leb128 encoded value to bigint.
+ * Decodes an unsigned leb128 encoded value to bigint. Note that if buffer
+ * that is provided does not _only_ contain the uleb128 encoded number an
+ * error will be thrown.
  *
  * @param {Buffer} buffer - The buffer to decode
  *
  * @returns {bigint} the decoded bigint value.
  */
 export const uleb128Decode = (buffer: Buffer): bigint => {
-    return buffer.reduce(
-        (result, byte, i) => result | (BigInt(byte & 0x7f) << BigInt(7 * i)), // For each byte, get the value of the 7 least significant bits (byte & 0x7f) and add this to the accumulator (<< 7 * i)
-        0n
-    );
+    const [bigint, index] = uleb128DecodeWithIndex(buffer);
+    if (index !== buffer.length + 1) {
+        throw 'The provided buffer did not contain just a single ULEB128 encoded number';
+    }
+    return bigint;
 };
 
+/**
+ * Decodes an unsigned leb128 encoded value to bigint and returns it along
+ * with the index of the end of the encoded uleb128 number + 1.
+ *
+ * @param {Buffer} bytes - The buffer to decode
+ * @param {number} index - The index where to decode at, defaults to 0
+ *
+ * @returns {[bigint, number]} the decoded bigint value and the index of
+ * the end of the encoded uleb128 number + 1.
+ */
+export function uleb128DecodeWithIndex(
```

Comment:
If you keep the index argument, make sure to state that it must be non-negative, since it is not uncommon to use negative numbers to offset from the end.